tortoise/organizer/organizer.go

```diff
@@ -0,0 +1,122 @@
+package organizer
+
+import (
+	"context"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// LayersIterator is used to traverse organized layers.
+// For siplicity LayersIterator iterates over all layers unconditionally.
+type LayersIterator func(types.LayerID)
+
+// Opt is for configuring Organizer.
+type Opt func(*Organizer)
+
+// WithLogger configures logger for Organizer.
+func WithLogger(logger log.Log) Opt {
+	return func(o *Organizer) {
+		o.logger = logger
+	}
+}
+
+// WithLastLayer layer configures last layer that tortoise knows about.
+func WithLastLayer(lid types.LayerID) Opt {
+	return func(o *Organizer) {
+		o.submitted = lid
+		o.received = lid
+	}
+}
+
+// WithWindowSize configures size of the window.
+// By default window of the size 1000 is used.
+func WithWindowSize(size int) Opt {
+	return func(o *Organizer) {
+		o.window = make([]types.LayerID, size)
+	}
+}
+
+// New creates Organizer instance.
+func New(opts ...Opt) *Organizer {
+	o := &Organizer{
+		logger: log.NewNop(),
+		window: make([]types.LayerID, 1000), // ~4kb
+	}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return o
+}
+
+// Organizer is a decorator for tortoise that keeps layers in order within the window
+// and sends them to tortoise only when the window has no gaps.
+type Organizer struct {
+	logger log.Log
+
+	// last submitted layer
+	submitted types.LayerID
+	// last received layer that is not submitted
+	received types.LayerID
+	// all layers between submitted and received
+	// window is non-empty only if there are gaps after submitted
+	window []types.LayerID
```

Comment:
can we call this buffer or something that's not `window`? as this can be easily confused with the sliding window in tortoise